sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample08_EventProcessor.md

```diff
@@ -0,0 +1,230 @@
+
+## What does a processor do?
+
+When consuming events, the consumer is tied to a specific event hub partition and reads the events in order from that partition. This means, to read all events from an Event Hub, you'll want to have one consumer per partition reading and processing events. For some workloads (where the processing can be intensive) you may want these consumers spread out across multiple machines, or even add or remove machines as the rate of incoming events increases or decreases. We call this process *load balancing*. In addition, you'll want to make sure that if one of your consumers has a problem (say the machine it is running on crashes) you're able to resume your processing at a later time, picking up where you left off. We call this process *checkpointing*.
+
+While you could write code directly using the `EventHubConsumerClient` to handle these cases, the Azure SDK also provides a type, `EventProcessorClient`, which handles this for you. It uses the Azure Blob Storage SDK to handle storing the state that's used during the load balancing and checkpointing operations. For many customers, that's sufficient, but some customers have asked for the ability to control how the information related to checkpointing and load balancing is stored. But for customers who have unique needs, we've exposed the lower-level machinery needed to build your own processor, which gives you full control over how this data is stored.
+
+## Building an Event Processor with `EventProcessor<TPartition>`
+
+The low level machinery needed to build you own processor, with full control over how its state is stored, is exposed by the `EventProcessor<TPartition>` class in the `Azure.Messaging.EventHubs.Primitives` namespace. This is an abstract class, which has five methods you'll need to provide implementations for. If you've used the `EventProcessorClient`, the first two will be familiar to you:
+
+- `OnProcessingEventBatchAsync`: The actual "business logic" of your processor. This is similar to the `ProcessEvent` event exposed by `EventProcessorClient`.
+- `OnProcessingErrorAsync`: A handler that can be used to observe exceptions from inside the machinery of the EventProcessor itself. This is similar to the `ProcessError` event exposed by `EventProcessorClient`.
+
+The remaining three deal with checkpointing and load balancing.
+
+- `ListOwnershipAsync`: Used by load balancing to see which partitions have been assigned to which processors.
+- `ClaimOwnershipAsync`: Used by load balancing to grab ownership of a partition (which may or may not be assigned to another processor).
+- `ListCheckpointsAsync`: Used to find the starting point inside a partition when a processor starts reading events from a partition.
```

Comment:
I'd strongly suggest talking about [`GetCheckpointAsync`](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor.cs#L806) and recommending that folks no-op `ListCheckpointsAsync` where possible.    That will allow them to return a specific checkpoint for the partition rather than the less efficient search through the collection on each iteration.

I have no idea why we originally implemented it as a "look through the list" but never cached it, but we recently added the optimization to help with performance.